pkg/v1/tkg/client/upgrade_cluster.go

```diff
@@ -1387,3 +1413,95 @@ func (c *TkgClient) configureOSOptionsForUpgrade(regionalClusterClient clustercl
+	newKCPPod, err := ModifyKubeVipTimeOutAndSerialize(&currentKubeVipPod, newLeaseDuration, newRenewDeadline, newRetryPeriod)
+	if err != nil {
+		return errors.Wrap(err, "unable to update kube-vip timeouts")
+	}
+
+	newKCP.Spec.KubeadmConfigSpec.Files[0].Content = newKCPPod
+
+	newKCPByte, err := json.Marshal(&newKCP)
+	if err != nil {
+		return errors.Wrap(err, "unable to marshal new KCP to byte array")
+	}
+	// Using polling to retry on any failed patch attempt. Sometimes if user upgrade
+	// workload cluster right after management cluster upgrade there is a chance
+	// that all controller pods are not started on management cluster
+	// and in this case patch fails. Retrying again should fix this issue.
+	pollOptions := &clusterclient.PollOptions{Interval: upgradePatchInterval, Timeout: upgradePatchTimeout}
+	err = regionalClusterClient.PatchResource(&capikubeadmv1beta1.KubeadmControlPlane{}, upgradeClusterConfig.KCPObjectName, upgradeClusterConfig.KCPObjectNamespace, string(newKCPByte), types.MergePatchType, pollOptions)
+	if err != nil {
+		return errors.Wrap(err, "unable to update the kubernetes version for kubeadm control plane nodes")
+	}
+
+	return nil
+}
+
+// ModifyKubeVipTimeOutAndSerialize modifies the time-out and lease duration parameters and serializes it to a string that can be patched
+func ModifyKubeVipTimeOutAndSerialize(currentKubeVipPod *corev1.Pod, newLeaseDuration, newRenewDeadline, newRetryPeriod string) (string, error) {
+	var envVars []corev1.EnvVar
+	for _, envVar := range currentKubeVipPod.Spec.Containers[0].Env {
+		if envVar.Name == vipLeaseDuration && envVar.Value == defaultLeaseDuration {
+			envVar.Value = newLeaseDuration
+		}
+		if envVar.Name == vipRenewDeadline && envVar.Value == defaultRenewDeadline {
+			envVar.Value = newRenewDeadline
+		}
+		if envVar.Name == vipRetryPeriod && envVar.Value == defaultRetryPeriod {
+			envVar.Value = newRetryPeriod
+		}
+		envVars = append(envVars, envVar)
+	}
+
+	currentKubeVipPod.Spec.Containers[0].Env = envVars
+
+	log.V(6).Infof("Marshaling kube-vip pod into a byte array")
+	kcpPodByte, err := yaml.Marshal(&currentKubeVipPod)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to marshal modified pod to byte array")
+	}
+
+	return string(kcpPodByte), nil
+}
+
+func (c *TkgClient) getNewKubeVipParameters() (string, string, string) {
+	newLeaseDuration := defaultNewLeaseDuration
+	newRenewDeadline := defaultNewRenewDeadline
+	newRetryPeriod := defaultNewRetryPeriod
+
+	leaseDuration, err := c.TKGConfigReaderWriter().Get(vipLeaseDurationKey)
+	if err == nil {
+		newLeaseDuration = leaseDuration
+	}
+
+	renewDeadline, err := c.TKGConfigReaderWriter().Get(vipRenewDeadlineKey)
+	if err == nil {
+		newRenewDeadline = renewDeadline
+	}
+
+	retryPeriod, err := c.TKGConfigReaderWriter().Get(vipRetryPeriodKey)
+	if err == nil {
+		newRetryPeriod = retryPeriod
+	}
+
+	return newLeaseDuration, newRenewDeadline, newRetryPeriod
+}
```

Comment:
Considering this might be a temporary hack and we would like to fix this as part of kubevip upgrade feature in the future if possible would be nice to keep these functions in a separate file (`kubevip_updates.go` or something similar). That way we can also define the parameter related constants to that separate file.